File: src/SwimReader.Server/wwwroot/src/MSAWVolume.js
// Ported 1:1 from _source/MSAWVolume.cs
// namespace DGScope

export class MSAWVolume {
    // [DisplayName("Name"), Category("Identity")]
    Name = null;    // string
    // [DisplayName("Active"), Category("Identity")]
    Active = true;  // bool
    // [Description("Draw the volume on the scope, for testing purposes."), Category("Identity")]
    Draw = false;   // bool
    // [DisplayName("Floor (ft MSL)"), Category("Geometry")]
    Floor = 0;      // int
    // [DisplayName("Ceiling / Minimum Safe Altitude (ft MSL)"), Category("Geometry")]
    Ceiling = 0;    // int
    // [DisplayName("Polygon Points"), Category("Geometry")]
    Points = [];    // List<GeoPoint>
    // Circle volumes: VolumeShape="Circle" with a Center and Radius > 0.
    // Polygon volumes leave Radius at 0 and use Points.
    // [DisplayName("Center (circle volumes)"), Category("Geometry")]
    Center = null;  // GeoPoint
    // [DisplayName("Radius (NM, 0 = polygon)"), Category("Geometry")]
    Radius = 0.0;   // double

    /// <summary>
    /// True if the given location is within this volume's horizontal footprint.
    /// </summary>
    ContainsLocation(location) { // bool ContainsLocation(GeoPoint location)
        if (location == null)
            return false;
        if (this.Radius > 0 && this.Center != null)
            return location.DistanceTo(this.Center) <= this.Radius;
        // Ray-casting point-in-polygon (handles the closing edge via the j=i++ wrap).
        if (this.Points == null || this.Points.length < 3)
            return false;
        let inside = false;
        let n = this.Points.length;
        for (let i = 0, j = n - 1; i < n; j = i++) {
            let xi = this.Points[i].Longitude, yi = this.Points[i].Latitude;
            let xj = this.Points[j].Longitude, yj = this.Points[j].Latitude;
            let intersect = ((yi > location.Latitude) != (yj > location.Latitude)) &&
                (location.Longitude < (xj - xi) * (location.Latitude - yi) / (yj - yi) + xi);
            if (intersect)
                inside = !inside;
        }
        return inside;
    }

    toString() {
        return this.Name;
    }
}

File: src/SwimReader.Server/Msaw/MSAWVolume.cs
using SwimReader.Server.Ca;

namespace SwimReader.Server.Msaw;

/// <summary>
/// Ported byte-faithful from DGScope's MSAWVolume.cs. A polygon or circle volume with a vertical
/// band [Floor, Ceiling); Ceiling doubles as the minimum safe altitude. Deserialized directly from
/// the profile XML's &lt;MSAWVolume&gt; elements (property names match DGScope).
/// </summary>
public class MSAWVolume
{
    public string? Name { get; set; }
    public bool Active { get; set; } = true;
    public bool Draw { get; set; }
    public int Floor { get; set; }
    public int Ceiling { get; set; }
    public List<GeoPoint> Points { get; set; } = new();
    public GeoPoint? Center { get; set; }
    public double Radius { get; set; }

    /// <summary>True if the given location is within this volume's horizontal footprint.</summary>
    public bool ContainsLocation(GeoPoint location)
    {
        if (location == null)
            return false;
        if (Radius > 0 && Center != null)
            return location.DistanceTo(Center) <= Radius;
        // Ray-casting point-in-polygon (handles the closing edge via the j=i++ wrap).
        if (Points == null || Points.Count < 3)
            return false;
        bool inside = false;
        int n = Points.Count;
        for (int i = 0, j = n - 1; i < n; j = i++)
        {
            double xi = Points[i].Longitude, yi = Points[i].Latitude;
            double xj = Points[j].Longitude, yj = Points[j].Latitude;
            bool intersect = ((yi > location.Latitude) != (yj > location.Latitude)) &&
                (location.Longitude < (xj - xi) * (location.Latitude - yi) / (yj - yi) + xi);
            if (intersect)
                inside = !inside;
        }
        return inside;
    }

    public override string ToString() => Name ?? "";
}
